api-search-googletranslate: Adds /api/v1/search, with translation query

antoviaque opened the pull request.

antoviaque committed:
api-search-googletranslate: Adds /api/v1/search, with translation query

talos commented on search/tests/api.py:
```diff
@@ -0,0 +1,139 @@
+            self.assertEqual(response_obj, reference)
+        except AssertionError:
+            expected = json.dumps(reference, indent=2, sort_keys=True)
+            message = "\nReturned:\n{returned}\n\nExpected:\n{expected}".format(returned=response.content,
```
Nitpicky, but since we're working with a lot of foreign languages and Python 2.x, can we make sure to use unicode strings in all cases?  Should we make a separate ticket for this, maybe?

antoviaque replied:
You're completely right, I really need to take the habit, especially in the context of the current project. Updated - don't hesitate to call me out on this whenever you see something I should do better in regards to encoding, or if I forget to use unicode anytime.

talos commented on search/tests/api.py:
```diff
@@ -0,0 +1,139 @@
+                }
+        mock_translator.translate.side_effect = mock_translate
+
+        response = self.client.get('/api/v1/search?expression=bom%20dia&query_type=translation')
```
À propos the above, I think a unit test with a non-ASCII character in the expression would be wise.

antoviaque replied:
Yup, added unicode characters in the request & answer.

talos commented on search/tests/api.py:
```diff
@@ -0,0 +1,139 @@
+
+        response = self.client.get('/api/v1/search?expression=bom%20dia&query_type=translation')
+        self.api_check(response, {
+                "error": 'test error one',
```
This strikes me as a little risky -- if we want the errors to be human-readable, spewing the text of the `Exception` through the API could be problematic.  For debug purposes it would be nice (perhaps through a separate property), but I'd rather we display a generic "Sorry! An error has occurred" unless it's a custom error subclass with a "description" property, perhaps.  This could look like this:

```
# We have an error object, 'exc'
if hasattr(error, u"description"):
    error_text = error.description
else:
    error_text = get_generic_error_description()
```

Alternatively, we could check the `Exception` to see if it's a languagelearning-specific subclass, then work from that.

Thoughts?

antoviaque replied:
Yes, as you noticed below, I've  put in place something similar to what you're describing for the custom error mechanism with the ErrorResponse exception: cf https://github.com/antoviaque/languagelearning/blob/api-search-googletranslate/utils/api.py#L26

In this particular case, the custom exception is used to display the raw exception message, as the googletranslate.py lib doesn't do a lot of error handling, so I figured we'd probably want to not crash the request if there was a bug there. But your comments made me realize this could be improved by doing this from the `dispatch()` method instead, to catch all exceptions without custom handling in the individual API methods. 

See if you like how I approached this: https://github.com/antoviaque/languagelearning/commit/498d462ed7b6dcefe47ff63cbc8d8feba54e96fb

talos commented on search/views/search.py:
```diff
@@ -0,0 +1,75 @@
+        self.source = request.GET.get('source', '').lower()
+        self.target = request.GET.get('target', 'en').lower()
+        self.query_types = request.GET.getlist('query_type', [])
+        # TODO: key, callback
```
:)

talos commented on search/views/search.py:
```diff
@@ -0,0 +1,75 @@
+        try:
+            response = translator.translate(self.expression, source=self.source, target=self.target)
+        except Exception as e:
+            raise ErrorResponse("; ".join(e.args))
```
This seems to be the relevant area for the `Exception` comment I made in the test above -- rather than joining the args of the error and spewing them directly over the API in all cases, it might be good to check to see if there's something specific we want to say if there's  an error description (and maybe do this if we're in DEBUG mode).

antoviaque replied:
Cf above comment, this has been removed - it will now either return the generic error when DEBUG is off, or let Django send back the full exception if DEBUG is on.

talos commented on search/views/search.py:
```diff
@@ -0,0 +1,75 @@
+        try:
+            translation = response['data']['translations'][0]
+            translated_text = translation['translatedText']
+        except KeyError:
```
I would do an explicit check for the key that is missing in case there is no translation -- it looks like it's probably `translation['translatedText']`?

```
if not u"translatedText" in translation:
    raise ErrorResponse('No translation found')
```

Otherwise we risk displaying "No translation found" for any class of `KeyError`, and making debugging more difficult.

antoviaque replied:
Idem - those cases should actually be handled by the `if 'error' in response` - if an unexpected structure is sent back, it will return the standard error message, as this should not happen without a bug in the googletranslate library or from the API answer.

talos commented on .gitignore:
```diff
@@ -5,3 +5,5 @@ languagelearning/local_settings.py
 *#
 node_modules
+var/cache/*
+var/db.sqlite3
```
The last file in this commit (`var/.empty`) is suspicious -- should it have been added here?

antoviaque replied:
`var/.empty` is there to make sure the var directory is created by git - that's the only way I know to force it when there are no actual files in the directory. Unless you know of a better way?

talos replied:
Ah, got it.  I don't know of a better way, although looking at this:

http://stackoverflow.com/questions/115983/how-do-i-add-an-empty-directory-to-a-git-repository

It seems that naming the file `.gitignore` is a slightly more "standard" way of hacking it… mainly, it makes it clear that the file is specifically git-related.

antoviaque replied:
Ah, good to know, thanks! Since I had merged the branch already, I've done the change on master: https://github.com/antoviaque/languagelearning/commit/aff972b4602963688c6f13fc33710a11758ae9bf

talos commented:
@antoviaque This looks good -- was able to pull and get tests running without a problem, thanks!  I raised some minor concerns about unicode, leaking error messages, and absorbing a `KeyError`.  Once those are integrated, we should be good to go!

Since none of those are blockers/change behavior, I'm going to start working on a feature branch based on this.  However, once those are fixed we should be good to merge into master, at which point I'll rebase from that.

antoviaque commented:
@talos Thanks for the comments, good catches! I'll go through those and fix as needed.

antoviaque committed:
api-search-googletranslate: Use unicode strings

talos commented on search/tests/api.py:
```diff
@@ -0,0 +1,140 @@
+            })
+        
+        # Translation throws an exception
+        def mock_translate(*args, **kwargs):
```
I didn't notice this before because I was reviewing pull request in GH instead of vim, but pyflakes doesn't care for the repeated redefinition of `mock_translate`.  I can't say I'm a huge fan, either.  Since these tests are essentially doing different things, would it be possible to break this one test case into three?  Any boilerplate setup could be abstracted into a separate method.

antoviaque replied:
Fixed in https://github.com/antoviaque/languagelearning/pull/6 - I've split the tests into smaller units.

talos commented on search/views/search.py:
```diff
@@ -0,0 +1,75 @@
+    def query_translation(self):
+        translator = GoogleTranslator()
+        try:
+            response = translator.translate(self.expression, source=self.source, target=self.target)
```
Having plugged this into the client, I've actually had extensive problems getting it to work with non-ASCII characters.  You should be able to reproduce such errors if you run a non-mocked version of the Google Translate test:

```
def test_api_search_translation_nonmocked(self):
    response = self.client.get(u'/api/v1/search?expression=bom%20dia%20e%C3%A8%C3%89%C9%98&query_type=translation')
    self.api_check(response, {
            u'expression': u'bom dia eèÉɘ',
            u'results': {
                u'translation': u'good day eèÉɘ'
            },
            u'source': u'pt',
            u'status': u'success',
            u'target': u'en'
        })
```

Yields

```
======================================================================
FAIL: test_api_search_translation_nonmocked (search.tests.api.SearchAPITest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/talos/Programming/languagelearning/search/tests/api.py", line 66, in test_api_search_translation_nonmocked        u'target': u'en'
  File "/Users/talos/Programming/languagelearning/search/tests/api.py", line 27, in api_check
    raise AssertionError(message)
AssertionError:
Returned:
{
  "error": "\u00e8",
  "expression": "bom dia e\u00e8\u00c9\u0258",
  "source": "",
  "status": "error",
  "target": "en"
}

Expected:
{
  "expression": "bom dia e\u00e8\u00c9\u0258",
  "results": {
    "translation": "good day e\u00e8\u00c9\u0258"
  },
  "source": "pt",
  "status": "success",
  "target": "en"
}
```

I fixed this in my client patch by turning `self.expression` into `self.expression.encode('utf-8')` so that the translator script is passed a bytestring instead of unicode, which is the type of `self.expression`.  However, this broke the mock tests.  I think some kind of non-mocked (integration) test to ensure the Google translate Python script can handle our non-ASCII inputs is necessary.

antoviaque replied:
Fixed in https://github.com/antoviaque/languagelearning/pull/6

I missed it because of how the googletranslate lib is mocked, I added a test mocking the actual http request rather than the full class to catch those. 

You got it right by adding the decode() call - the lib doesn't handle unicode characters (quote() expects bytestrings). This broke the mock tests because tests control the value passed to the translate method, and those aare changed when the string is encoded to utf8.

antoviaque committed:
api-search-googletranslate: Improved API error handling

* Exceptions thrown while processing an API request now return the
  proper JSON formatted error response without the need for custom
  handling. If `DEBUG=True`, this behavior is disabled and Django
  will return the full traceback.
* Added support for custom status code (and any argument for
  HttpResponse) to `APIView.render_to_response`
* Added check for status code to `api_check()`

Note - the changes are backward-incompatible.

antoviaque commented:
Done!

When you rebase, have a look at https://github.com/antoviaque/languagelearning/commit/498d462ed7b6dcefe47ff63cbc8d8feba54e96fb to take into account the backward-incompatible changes to the error responses and `api_check()`.

(And actually, you can probably merge rather than rebase here too, to not lose the history in such a situation either, which would show the fact that you originally wrote code against the previous behavior, and had to adapt it to the new.)

talos commented on search/views/search.py:
```diff
@@ -0,0 +1,70 @@
+    
+    def query_translation(self):
+        translator = GoogleTranslator()
+        response = translator.translate(self.expression, source=self.source, target=self.target)
```
Huh, maybe I forgot to press "comment on this line", but GH seems to have eaten a comment I entered earlier this morning.  I've had problems with this in the client branch off of this pull.  In particular, a non-mocked test case:

```
def test_api_search_translation_nomock(self):
    response = self.client.get(u'/api/v1/search?expression=bom%20dia%20e%C3%A8%C3%89%C9%98&query_type=translation')
    self.api_check(response, 200, {
            u'expression': u'bom dia eèÉɘ',
            u'results': {
                u'translation': u'good day eèÉɘ'
            },
            u'source': u'pt',
            u'status': u'success',
            u'target': u'en'
        })
```

Will fail:

```
Traceback (most recent call last):
  File "/Users/talos/Programming/languagelearning/search/tests/api.py", line 75, in test_api_search_translation_nomock
    u'target': u'en'
  File "/Users/talos/Programming/languagelearning/search/tests/api.py", line 28, in api_check
    self.assertEqual(response.status_code, status_code)
AssertionError: 400 != 200
```

This can be solved by encoding `self.expression` into a UTF-8 bytestring for the `GoogleTranslator` (`self.expression.encode('utf8')`), but this seems a bit messy and also breaks the mocked test cases.  Is there something else that I'm missing in terms of addressing this?

antoviaque commented:
And done for logging of exceptions & the #TODO LOCAL on https://github.com/antoviaque/languagelearning/pull/6 at https://github.com/antoviaque/languagelearning/commit/d3325f4dc0365b7f53de8b075a09959072540531
